### PR TITLE
Contract Execution Transaction Specification

### PR DESCRIPTION
**Pull request by nkohen**
Took my best shot at specifying the DLC Contract Execution Transaction as things stand right now.

The CET weight has been tested here: https://blockstream.info/testnet/tx/e2678c5646dc79ed0d7e0cc79ef5b0b3539d8ec9e32de1fbb3d854aaab1709f6

Related to #12 
Related to #46

TODO:
- [ ] CET Test Vectors

### CODE REVIEW

File: Transaction/DLC_Contract_Execution_Transaction.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+There will be one CET for every possible outcome, the output values correspond to such an outcome and are negotiated in the offer message.
+
+## CET Outputs
+
+The Offerer/Initiator's output comes first, the Accepter's output comes second.
+
+Note that this will likely change in the future.
+
+If either party receives less than the dust limit of `1000 satoshis` for this outcome, then their output is not produced.
```
Comment (benthecarman) [inline Transaction/DLC_Contract_Execution_Transaction.md:36]: Should note that both of these are likely to change

Comment (benthecarman) [reply Transaction/DLC_Contract_Execution_Transaction.md:36]: Also, I think this is another good case for where we could use bip-69 for input/output ordering for v0.1

**Comment by nkohen** [comment]
I'm realizing that the `Refund Transaction` is exactly the same as a `CET` except it has a different locktime and output values are equal to the offerer and accepter's total collaterals from contract negotiation. As such, I am no starting to lean towards once again having just a single specification document for transactions which combines this PR and #51 into a single file and has a section on the Refund Transaction stating that it is specified as a special `CET` construction so that there isn't so much redundancy.

File: Transaction/DLC_Contract_Execution_Transaction.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+multi_sig: 71 bytes
+	- OP_2: 1 byte
+	- OP_DATA: 1 byte (offer_pubkey length)
+	- offer_pubkey: 33 bytes
+	- OP_DATA: 1 byte (accept_pubkey length)
+	- accept_pubkey: 33 bytes
+	- OP_2: 1 byte
+	- OP_CHECKMULTISIG: 1 byte
+
+offer_output: 9 + offer_output_script length bytes
```
Comment (ariard) [inline Transaction/DLC_Contract_Execution_Transaction.md:60]: Aren't output scripts normalized and thus we know their length ?

File: Transaction/DLC_Contract_Execution_Transaction.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+If either party receives less than the dust limit of `1000 satoshis` for this outcome, then their output is not produced.
+
+### Offerer Output
+
+This output sends funds won by the offerer corresponding to this CET's outcome to the offerer's final address specified in the offer message.
+
+### Accepter Output
+
+This output sends funds won by the accepter corresponding to this CET's outcome to the accepter's final address specified in the accept message.
```
Comment (ariard) [inline Transaction/DLC_Contract_Execution_Transaction.md:44]: Replace address by scriptpubkey. Address are application-level data structs.

File: Transaction/DLC_Contract_Execution_Transaction.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+The Offerer/Initiator's output comes first, the Accepter's output comes second.
+
+Note that this will likely change in the future.
+
+If either party receives less than the dust limit of `1000 satoshis` for this outcome, then their output is not produced.
+
+### Offerer Output
+
+This output sends funds won by the offerer corresponding to this CET's outcome to the offerer's final address specified in the offer message.
```
Comment (ariard) [inline Transaction/DLC_Contract_Execution_Transaction.md:40]: Layout the witnessScript as they're not P2WPKH IIRC?

File: Transaction/DLC_Contract_Execution_Transaction.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+### Accepter Output
+
+This output sends funds won by the accepter corresponding to this CET's outcome to the accepter's final address specified in the accept message.
+
+# Fees
+
+Fees for the CET are paid for in the Funding Transaction, the following only shows the weight calculation for the largest possible CET.
```
Comment (ariard) [inline Transaction/DLC_Contract_Execution_Transaction.md:48]: You mean the feerate is the same than the funding transaction ? Might be better like "The fee calculation for CETs is based on the negotiated feerate_per_kw and the expected weight of the transaction."

**Review by ariard — COMMENTED** [review]
I think we should discuss if we add CET anchor outputs for v0.1. IMO, as predicting the feerate is even harder to be right for CET given your oracle signatures might be a month from now, network mempools will be far different. With the funding, the feerate negotiated is to be broadcast soon so less likely to be inaccurate.

**Comment by nkohen** [comment]
Replaced by #57

File: Transaction/DLC_Contract_Execution_Transaction.md (hunk 2 above)
Comment (nkohen) [reply Transaction/DLC_Contract_Execution_Transaction.md:60]: No, there is a final address field in both the offer and accept messages which will specify the SPK used, it does not even have to be segwit.

File: Transaction/DLC_Contract_Execution_Transaction.md (hunk 3 above)
Comment (nkohen) [reply Transaction/DLC_Contract_Execution_Transaction.md:44]: addresses are what are currently specified in the offer and accept messages.

File: Transaction/DLC_Contract_Execution_Transaction.md (hunk 2 above)
Comment (benthecarman) [reply Transaction/DLC_Contract_Execution_Transaction.md:60]: p2wpkh is `OP_0 <20-byte-hash>` vs p2wsh is `OP_0 <32-byte-hash>`

taproot will be `OP_1 <32-byte-pub-key>`

File: Transaction/DLC_Contract_Execution_Transaction.md (hunk 4 above)
Comment (nkohen) [reply Transaction/DLC_Contract_Execution_Transaction.md:40]: not sure what you mean?

File: Transaction/DLC_Contract_Execution_Transaction.md (hunk 5 above)
Comment (nkohen) [reply Transaction/DLC_Contract_Execution_Transaction.md:48]: No, I mean that the outputs of the CETs are the exact outcome values negotiated and this is accomplished by including their fees in the funding output

Comment (nkohen) [reply Transaction/DLC_Contract_Execution_Transaction.md:48]: I.e. all fee computation is done when constructing the funding transaction, none is necessary for the CETs/Refund Tx

File: Transaction/DLC_Contract_Execution_Transaction.md (hunk 4 above)
Comment (ariard) [reply Transaction/DLC_Contract_Execution_Transaction.md:40]: Gotcha, they're freely picked up by offerer/accepter ?

**Pull request closed without merging**